BitFaster.Caching/Lru/TemplateConcurrentLru.cs

```diff
@@ -295,24 +295,110 @@ public void AddOrUpdate(K key, V value)
         ///<inheritdoc/>
         public void Clear()
         {
-            // take a key snapshot
-            var keys = this.dictionary.Keys.ToList();
+            int count= this.Count();
 
-            // remove all keys in the snapshot - this correctly handles disposable values
-            foreach (var key in keys)
+            for (int i = 0; i < count; i++)
             {
-                TryRemove(key);
+                CycleHotUnchecked(ItemRemovedReason.Cleared);
+                CycleWarmUnchecked(ItemRemovedReason.Cleared);
+                CycleColdUnchecked(ItemRemovedReason.Cleared);
             }
+        }
+
+        /// <summary>
+        /// Trim the specified number of items from the cache. Removes all discardable items per IItemPolicy.ShouldDiscard(), then 
+        /// itemCount-discarded items in LRU order, if any.
+        /// </summary>
+        /// <param name="itemCount">The number of items to remove.</param>
+        /// <returns>The number of items removed from the cache.</returns>
+        /// <exception cref="ArgumentOutOfRangeException"><paramref name="itemCount"/> is less than 0./</exception>
+        /// <exception cref="ArgumentOutOfRangeException"><paramref name="itemCount"/> is greater than capacity./</exception>
+        /// <remarks>
+        /// Note: Trim affects LRU order. Calling Trim resets the internal accessed status of items.
+        /// </remarks>
+        public void Trim(int itemCount)
+        {
+            int capacity = this.coldCapacity + this.warmCapacity + this.hotCapacity;
+
+            if (itemCount < 0 || itemCount > capacity)
+            { 
+                throw new ArgumentOutOfRangeException(nameof(itemCount), "itemCount must be greater than or equal to zero, and less than the capacity of the cache.");
+            }
+
+            // clamp itemCount to number of items actually in the cache
+            itemCount = Math.Min(itemCount, this.HotCount + this.WarmCount + this.ColdCount);
+
+            // first scan each queue for discardable items and remove them immediately. Note this can remove > itemCount items.
+            int itemsRemoved = TrimAllDiscardedItems();

```

Comment:
This is always O(n), which is wasteful for the LRU case that will never discard any items.